### PR TITLE
FieldConfig: add thresholds and color modes

### PR DESCRIPTION
**Pull request by ryantxu**
Alternative to #20783

This changes the idea of Scale into:

```
export enum ThresholdsMode {
  Absolute = 'absolute',
  Percentage = 'percentage', // between 0 and 1 (based on min/max)
}

export interface Thresholds {
  mode: ThresholdsMode;

  // Must be sorted by 'value', first value is always -Infinity
  step: Threshold[];
}
```

And:
```
export enum FieldColorMode {
  Thresholds = 'thresholds',
  Scheme = 'scheme',
  Fixed = 'fixed',
}

export interface FieldColor {
  mode: FieldColorMode;
  schemeName?: ColorScheme;
  fixedColor?: string;
}
```

### CODE REVIEW

**Review by torkelo — APPROVED** [review]
Think this is looking solid now, left some minor comments where I got a bit confused. And some ideas around type & property names.

File: packages/grafana-data/src/types/thresholds.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+export enum ThresholdsMode {
+  Absolute = 'absolute',
+  Percentage = 'percentage', // between 0 and 1 (based on min/max)
+}
+
+export interface Thresholds {
+  mode: ThresholdsMode;
+
+  // Must be sorted by 'value', first value is always -Infinity
+  step: Threshold[];
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply packages/grafana-data/src/types/thresholds.ts]: 👍

File: packages/grafana-data/src/types/dataFrame.ts [hunk 2]
```diff
@@ -32,8 +33,11 @@ export interface FieldConfig {
   // Convert input values into a display string
   mappings?: ValueMapping[];
 
-  // Must be sorted by 'value', first value is always -Infinity
-  thresholds?: Threshold[];
+  // Map numeric values to states
+  thresholds?: Thresholds;
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply packages/grafana-data/src/types/dataFrame.ts]: changed to ThresholdsConfig and used `steps` where appropriate

File: packages/grafana-ui/src/components/BarGauge/BarGauge.tsx [hunk 3]
```diff
@@ -495,26 +502,54 @@ export function getBasicAndGradientStyles(props: Props): BasicAndGradientStyles
...
+      const pos = valuePercent * maxSize;
+      const offset = Math.round(pos - (pos - lastpos) / 2);
+
+      if (gradient === '') {
+        gradient = `linear-gradient(${cssDirection}, ${color}, ${color}`;
+      } else {
+        lastpos = pos;
+        gradient += ` ${offset}px, ${color}`;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply packages/grafana-ui/src/components/BarGauge/BarGauge.tsx]: I renamed the variables so they are more specific to the color schemes.

File: packages/grafana-ui/src/components/ThresholdsEditor/ThresholdsEditor.tsx [hunk 4]
```diff
@@ -40,10 +51,30 @@ export class ThresholdsEditor extends PureComponent<Props, State> {
...
+          ],
+        },
+        () => this.onChange()
+      );
+    } else if (!thresholds.mode) {
+      this.props.onChange({
+        ...thresholds,
+        mode: ThresholdsMode.Absolute,
+      });
+    }
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply packages/grafana-ui/src/components/ThresholdsEditor/ThresholdsEditor.tsx]: got rid of it -- I was testing with a panel saved wrong and was using this to reset -- but as you say, that should not happen for real

**Pull request merged**